### DISCUSSION TITLE
Receiving "TypeError: cannot pickle '_thread.lock' object" with providers.Factory

### DISCUSSION THREAD

**Discussion by chriscoindeskindices** [ets-labs/python-dependency-injector#718] [Q&A]
I am using providers.Factory to create a service, and one of the dependencies is a class to interact with our Redis database.

This line in the code seems to be the offender:
`self._client = redis.Redis(host=host, port=port, password=password, db=db, health_check_interval=health_check_interval)`

that code is executing in the constructor.  if I remove that line or if I change it to 
`client = redis.Redis(host=host, port=port, password=password, db=db, health_check_interval=health_check_interval)`

which doesn't help me but just an observation--then there is no error.  I understand what pickling is, but I don't understand what is going on here exactly.

**Comment by xOctave** [comment]
You cannot pass the Redis object directly (or whatever other object). Instead, you need to wrap it with a `providers.Object(.)`. Refer to the [documentation](https://python-dependency-injector.ets-labs.org/providers/object.html).

It would be something like this

```
# import are here to avoid ambiguity
from dependency_injector import providers

client = providers.Object(redis.Redis(host=host, ...))
```

Hope it helps.
